diff --git a/src/argprocessing.cpp b/src/argprocessing.cpp
index 0bc37e57..75b7cfa2 100644
--- a/src/argprocessing.cpp
+++ b/src/argprocessing.cpp
@@ -287,7 +287,7 @@ process_option_arg(const Context& ctx,
   }
 
   bool changed_from_slash = false;
-  if (ctx.config.is_compiler_group_msvc() && util::starts_with(args[i], "/")) {
+  if (ctx.config.is_compiler_group_msvc() && !state.found_c_opt && util::starts_with(args[i], "/")) {
     // MSVC understands both /option and -option, so convert all /option to
     // -option to simplify our handling.
     args[i][0] = '-';
